src/upgrade.rs

```diff
@@ -317,26 +317,34 @@ mod sealed {
     }
 
     impl CanUpgrade for http::Request<crate::Body> {
-        fn on_upgrade(self) -> OnUpgrade {
-            self.into_body().take_upgrade()
+        fn on_upgrade(mut self) -> OnUpgrade {
+            self.extensions_mut()
+                .remove::<OnUpgrade>()
+                .unwrap_or(OnUpgrade::none())
```

Comment:
You can save a function call by writing this as `.unwrap_or_else(OnUpgrade::none)` (and in the lines below).